File: test/com/routon/os/TResult.java
/*
 * Author:lijin@routon
 * Date:20121225
 */

package com.routon.os;

import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;

public class TResult implements Parcelable {
    private int mStatus = 0;
    private String mStdoutStr = null;
    private String mStderrStr = null;

    public TResult(int rc, String stdout, String stderr) {
        mStatus = rc;
        mStdoutStr = stdout;
        mStderrStr = stderr;
    }

    public TResult(Parcel in) {
        mStatus = in.readInt();
        mStdoutStr = in.readString();
        mStderrStr = in.readString();
    }

    public int getStatus() {
        return mStatus;
    }

    public String getStdoutString() {
        return mStdoutStr;
    }

    public String getStderrString() {
        return mStderrStr;
    }

    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel out, int flags) {
        out.writeInt(mStatus);
        out.writeString(mStdoutStr);
        out.writeString(mStderrStr);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("TResult: mStatus("); sb.append(mStatus); sb.append(") ");
        sb.append("mStdoutStr("); sb.append(mStdoutStr); sb.append(") ");
        sb.append("mStderrStr("); sb.append(mStderrStr); sb.append(") ");
        return sb.toString();
    }

    public static final Parcelable.Creator<TResult> CREATOR =
            new Parcelable.Creator<TResult>() {
        public TResult createFromParcel(Parcel in) {
            return new TResult(in);
        }
        public TResult[] newArray(int size) {
            return new TResult[size];
        }
    };
}

File: test/tprocess/TResult.cpp
#include "TResult.h"
#include <utils/TextOutput.h>

namespace com {
    namespace routon {
        namespace os {
            TResult::~TResult() {
                android::aerr << "TResult::~TResult"<<android::endl;
            }
            void TResult::writeToParcel(Parcel *out, int flags) {
                out->writeInt32(mStatus);
                out->writeString8(mStdoutStr);
                out->writeString8(mStderrStr);
                /*fprintf(stderr, " writeToParcel  %d out %s err %s\n", */
                /*mStatus, String8(mStdoutStr).string(),*/
                /*String8(mStderrStr).string());*/

            }

            int TResult::describeContents() {
                return 0;
            }

            TResult *TResult::createFromParcel(const Parcel& source){
                String8 err = source.readString8();
                String8 ou = source.readString8();
                int st = source.readInt32();
                return new TResult(st,ou, err);
                //return NULL;
            }

            void TResult::readFromParcel(const Parcel& source) {
                this->mStderrStr =  source.readString8();
                this->mStdoutStr =  source.readString8();
                this->mStatus =  source.readInt32();
            }

            std::vector<Parcelable *> TResult::newArray(int size) {
                std::vector<Parcelable *> v;
                for (int i = 0; i < size; ++i) {
                    v.push_back(new TResult(0,String8(""),String8("")));
                }

                return v;
            }
        };
    };
};
